sparse-traits/examples/playground.rs

```diff
@@ -8,34 +8,59 @@ struct OpWithMatVec {}
 #[derive(Debug)]
 struct OpWithoutMatVec {}
 
+struct SimpleSpace;
+
+impl LinearSpace for SimpleSpace {
+    type Item = f64;
+    type Real = f64;
+    type VectorType = Vec;
+}
+
+struct View;
+impl VectorView for View {
+    type Item = f64;
+}
+
 // Simple helper structs as mock vectors
 #[derive(Debug)]
 struct Vec {}
-impl Vector for Vec {}
+impl Vector for Vec {
+    type Space = SimpleSpace;
+    type View = View;
+}
 
 // For OpWithMatvec we have to implement
 // `as_matvec` to return a reference to self.
 // This is just boilerplate and could be handled
 // by a simple derive macro or similar so the user
 // needs not write this boilerplate.
 impl OperatorBase for OpWithMatVec {
-    fn as_matvec(&self) -> Option<&dyn AsMatVec> {
+    type In = SimpleSpace;
+    type Out = SimpleSpace;
+    fn as_matvec(&self) -> Option<&dyn AsMatVec<In = Self::In, Out = Self::Out>> {
         Some(self)
     }
 }
 
 // The actual matvec is now implemented. It is just
 // a stub that prints a message.
 impl AsMatVec for OpWithMatVec {
-    fn matvec(&self, _x: &dyn Vector, _y: &mut dyn Vector) -> Result<(), Error> {
+    fn matvec(
+        &self,
+        _x: &<Self::In as LinearSpace>::VectorType,
+        _y: &mut <Self::Out as LinearSpace>::VectorType,
```

Comment:
I think this interface is like taking an `(&String, &mut String)`, but we don't need references to owned things. If we `impl AsRef<VectorView>`, then the operators will only need a view (which could be a sub-vector of the original).

Reply:
I think this won't work. If we do this in the operator base we cannot have them as trait classes any more since `impl AsRef<VectorView>` is a generic.

Reply:
Why not `<Self::In as LinearSpace>::VectorType::VectorViewType` or (I think more idiomatically) using `Deref` just like how `String` derefs to `str`?